src/line/i-chartBuilder.test.js

```diff
@@ -0,0 +1,400 @@
+/**
+ * @jest-environment jsdom
+ */
+
+const fs = require("fs")
+const domTesting = require('@testing-library/dom')
+require('@testing-library/jest-dom')
+const userEvent = require("@testing-library/user-event").default
+
+//Spy on
+const window = require("./window")
+ 
+function initDomFromFiles(htmlPath, jsPath) {
+    window.localStorage.clear()
+    const html = fs.readFileSync(htmlPath, 'utf8')
+    document.open()
+    document.write(html)
+    document.close()
+    jest.isolateModules(function() {
+        require(jsPath)
+    })
+}
+
+/*
+    They should verify the behaviors described above. Note that these behaviors are 
+    duplicated across the three chart builder pages (bar.html, line.html, and scatter.html). 
+    You only need to test these behaviors on one of those pages.
+*/
+
+//ADDING VALUES TO THE CHART BUILDER
+
+test('Pressing "+" adds exactly one empty x and y value box', async function(){
+    initDomFromFiles(`${__dirname}/line.html`, `${__dirname}/line.js`)
+
+    //get + button for adding  new x and y cells.
+    const buttons = domTesting.queryAllByRole(document, "button")
+    const addButton = buttons[3];
+    
+    //make sure the correct button is selected
+    expect(addButton).toHaveTextContent("+")
+
+    //count all cells
+    var xValues = await domTesting.getAllByLabelText(document, "X")
+    var yValues = await domTesting.getAllByLabelText(document, "Y")
+
+
+    // Assert that the lists are not empty
+	expect(xValues).toHaveLength(1)
+    expect(yValues ).toHaveLength(1)
+
+    const user = userEvent.setup()
+
+    await user.click(addButton)
+
+    xValues = domTesting.getAllByLabelText(document, "X")
+    yValues = domTesting.getAllByLabelText(document, "Y")
+
+    expect(xValues).toHaveLength(2)
+    expect(yValues).toHaveLength(2)
+})
+
+
+//ALERTS DISPLAYED FOR MISSING CHART DATA
+test('Error message appears for missing chart title', async function(){
+    initDomFromFiles(`${__dirname}/line.html`, `${__dirname}/line.js`)
+
+    const xLabel = domTesting.getByLabelText(document, "X label")
+    const yLabel = domTesting.getByLabelText(document, "Y label")
+    
+    const xValues = domTesting.getAllByLabelText(document, "X")
+    const yValues = domTesting.getAllByLabelText(document, "Y")
+    const xValue = xValues[0]
+    const yValue = yValues[0]
+
+    const buttons = domTesting.queryAllByRole(document, "button")
+    var generateButton = buttons[0]
+
+    //make sure the correct button is selected
+    expect(generateButton).toHaveTextContent("Generate chart")
+
+    //Popups
+    window.alert = jest.fn();
```

Comment:
Not sure it matters a whole ton, but does `window.alert` need to be mocked? I would've thought that `jsdom` would already provide a suitable implementation of the function, and it adds an extra mock to clean up between tests.